File: galpy/potential/LogarithmicHaloPotential.py
###############################################################################
#   LogarithmicHaloPotential.py: class that implements the logarithmic
#                            potential Phi(r) = vc**2 ln(r)
###############################################################################
import warnings

import numpy

from ..util import conversion, galpyWarning
from .Potential import Potential, kms_to_kpcGyrDecorator

_CORE = 10**-8


class LogarithmicHaloPotential(Potential):
    """Class that implements the logarithmic potential

    .. math::

        \\Phi(R,z) = \\frac{\\mathrm{amp}}{2}\\,\\ln\\left[R^2+\\left(\\frac{z}{q}\\right)^2+\\mathrm{core}^2\\right]

    Alternatively, the potential can be made triaxial by adding a parameter :math:`b`

    .. math::

        \\Phi(x,y,z) = \\frac{\\mathrm{amp}}{2}\\,\\ln\\left[x^2+\\left(\\frac{y}{b}\\right)^2+\\left(\\frac{z}{q}\\right)^2+\\mathrm{core}^2\\right]

    With these definitions, :math:`\\sqrt{\\mathrm{amp}}` is the circular velocity at :math:`r \\gg \\mathrm{core}` at :math:`(y,z) = (0,0)`.

    """

    def __init__(
        self, amp=1.0, core=_CORE, q=1.0, b=None, normalize=False, ro=None, vo=None
    ):
        """
        Initialize a logarithmic potential.

        Parameters
        ----------
        amp : float or Quantity, optional
            Amplitude to be applied to the potential; can be a Quantity with units of velocity-squared.
        core : float or Quantity, optional
            Core radius at which the logarithm is cut.
        q : float
            Potential flattening (z/q)**2.
        b : float, optional
            Shape parameter in y-direction (y --> y/b; see definition).
        normalize : bool or float, optional
            If True, normalize such that vc(1.,0.)=1., or, if given as a number, such that the force is this fraction of the force necessary to make vc(1.,0.)=1.
        ro : float, optional
            Distance scale for translation into internal units (default from configuration file).
        vo : float, optional
            Velocity scale for translation into internal units (default from configuration file).

        Notes
        -----
        - 2010-04-02 - Started - Bovy (NYU)
        """
        Potential.__init__(self, amp=amp, ro=ro, vo=vo, amp_units="velocity2")
        core = conversion.parse_length(core, ro=self._ro)
        self.hasC = True
        self.hasC_dxdv = True
        self.hasC_dens = True
        self._core2 = core**2.0
        self._q = q
        self._b = b
        if not self._b is None:
            self.isNonAxi = True
            self._1m1overb2 = 1.0 - 1.0 / self._b**2.0
        if normalize or (
            isinstance(normalize, (int, float)) and not isinstance(normalize, bool)
        ):  # pragma: no cover
            self.normalize(normalize)
        self._nemo_accname = "LogPot"
        return None

    def _evaluate(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            return (
                1.0
                / 2.0
                * numpy.log(
                    R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
                    + (z / self._q) ** 2.0
                    + self._core2
                )
            )
        else:
            return 1.0 / 2.0 * numpy.log(R**2.0 + (z / self._q) ** 2.0 + self._core2)

    def _Rforce(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            return -Rt2 / R / (Rt2 + (z / self._q) ** 2.0 + self._core2)
        else:
            return -R / (R**2.0 + (z / self._q) ** 2.0 + self._core2)

    def _zforce(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            return -z / self._q**2.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
        else:
            return -z / self._q**2.0 / (R**2.0 + (z / self._q) ** 2.0 + self._core2)

    def _phitorque(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            return (
                R**2.0
                / (Rt2 + (z / self._q) ** 2.0 + self._core2)
                * numpy.sin(2.0 * phi)
                * self._1m1overb2
                / 2.0
            )
        else:
            return 0

    def _dens(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            R2 = R**2.0
            Rt2 = R2 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            denom = 1.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
            denom2 = denom**2.0
            return (
                1.0
                / 4.0
                / numpy.pi
                * (
                    2.0 * Rt2 / R2 * (denom - Rt2 * denom2)
                    + denom / self._q**2.0
                    - 2.0 * z**2.0 * denom2 / self._q**4.0
                    - self._1m1overb2
                    * (
                        2.0
                        * R2
                        * numpy.sin(2.0 * phi) ** 2.0
                        / 4.0
                        * self._1m1overb2
                        * denom2
                        + denom * numpy.cos(2.0 * phi)
                    )
                )
            )
        else:
            return (
                1.0
                / 4.0
                / numpy.pi
                / self._q**2.0
                * (
                    (2.0 * self._q**2.0 + 1.0) * self._core2
                    + R**2.0
                    + (2.0 - self._q**-2.0) * z**2.0
                )
                / (R**2.0 + (z / self._q) ** 2.0 + self._core2) ** 2.0
            )

    def _R2deriv(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            denom = 1.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
            return (denom - 2.0 * Rt2 * denom**2.0) * Rt2 / R**2.0
        else:
            denom = 1.0 / (R**2.0 + (z / self._q) ** 2.0 + self._core2)
            return denom - 2.0 * R**2.0 * denom**2.0

    def _z2deriv(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            denom = 1.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
            return denom / self._q**2.0 - 2.0 * z**2.0 * denom**2.0 / self._q**4.0
        else:
            denom = 1.0 / (R**2.0 + (z / self._q) ** 2.0 + self._core2)
            return denom / self._q**2.0 - 2.0 * z**2.0 * denom**2.0 / self._q**4.0

    def _Rzderiv(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            return (
                -2.0
                * Rt2
                / R
                * z
                / self._q**2.0
                / (Rt2 + (z / self._q) ** 2.0 + self._core2) ** 2.0
            )
        else:
            return (
                -2.0
                * R
                * z
                / self._q**2.0
                / (R**2.0 + (z / self._q) ** 2.0 + self._core2) ** 2.0
            )

    def _phi2deriv(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            denom = 1.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
            return -self._1m1overb2 * (
                R**4.0
                * numpy.sin(2.0 * phi) ** 2.0
                / 2.0
                * self._1m1overb2
                * denom**2.0
                + R**2.0 * denom * numpy.cos(2.0 * phi)
            )
        else:
            return 0.0

    def _Rphideriv(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            denom = 1.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
            return (
                -(denom - Rt2 * denom**2.0) * R * numpy.sin(2.0 * phi) * self._1m1overb2
            )
        else:
            return 0.0

    def _phizderiv(self, R, z, phi=0.0, t=0.0):
        if self.isNonAxi:
            Rt2 = R**2.0 * (1.0 - self._1m1overb2 * numpy.sin(phi) ** 2.0)
            denom = 1.0 / (Rt2 + (z / self._q) ** 2.0 + self._core2)
            return (
                2
                * R**2
                * z
                * numpy.sin(phi)
                * numpy.cos(phi)
                * self._1m1overb2
                * denom**2
                / self._q**2
            )
        else:
            return 0.0

    @kms_to_kpcGyrDecorator
    def _nemo_accpars(self, vo, ro):
        warnings.warn(
            "NEMO's LogPot does not allow flattening in z (for some reason); therefore, flip y and z in NEMO wrt galpy; also does not allow the triaxial b parameter",
            galpyWarning,
        )
        ampl = self._amp * vo**2.0
        return "0,{},{},1.0,{}".format(
            ampl,
            self._core2
            * ro**2.0
            * self._q ** (2.0 / 3.0),  # somewhat weird gyrfalcon implementation
            self._q,
        )

File: galpy/potential/potential_c_ext/LogarithmicHaloPotential.c
#include <math.h>
#include <galpy_potentials.h>
//LogarithmicHaloPotential
//4 (3)  arguments: amp, c2, (and q), now also 1-1/b^2 for triaxial!
double LogarithmicHaloPotentialEval(double R,double Z, double phi,
				    double t,
				    struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double q= *(args+1);
  double c= *(args+2);
  double onem1overb2= *(args+3);
  //Calculate potential
  double zq= Z/q;
  if ( onem1overb2 < 1 )
    return 0.5 * amp * log(R*R * (1. - onem1overb2 * pow(sin(phi),2))+zq*zq+c);
  else
    return 0.5 * amp * log(R*R+zq*zq+c);
}
double LogarithmicHaloPotentialRforce(double R,double Z, double phi,
				      double t,
				      struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double q= *(args+1);
  double c= *(args+2);
  double onem1overb2= *(args+3);
  //Calculate Rforce
  double zq= Z/q;
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return - amp * Rt2/R/(Rt2+zq*zq+c);
  } else
    return - amp * R/(R*R+zq*zq+c);
}
double LogarithmicHaloPotentialPlanarRforce(double R,double phi,
					    double t,
					    struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double c= *(args+2); // skip q
  double onem1overb2= *(args+3);
  //Calculate Rforce
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return -amp * Rt2/R/(Rt2+c);
  } else
    return -amp * R/(R*R+c);
}
double LogarithmicHaloPotentialzforce(double R,double z,double phi,
				      double t,
				      struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double q= *(args+1);
  double c= *(args+2);
  double onem1overb2= *(args+3);
  //Calculate zforce
  double zq= z/q;
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return -amp * zq/q/(Rt2+zq*zq+c);
  } else
    return -amp * zq/q/(R*R+zq*zq+c);
}
double LogarithmicHaloPotentialphitorque(double R,double z,double phi,
					double t,
					struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double q= *(args+1);
  double c= *(args+2);
  double onem1overb2= *(args+3);
  //Calculate phitorque
  double zq;
  double Rt2;
  if ( onem1overb2 < 1 ) {
    zq= z/q;
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return amp * R*R / (Rt2+zq*zq+c) * sin(2*phi) * onem1overb2 / 2.;
  } else
    return 0.;
}
double LogarithmicHaloPotentialPlanarphitorque(double R,double phi,
					      double t,
					      struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double c= *(args+2); // skip q
  double onem1overb2= *(args+3);
  //Calculate phitorque
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return amp * R*R / (Rt2+c) * sin(2*phi) * onem1overb2 / 2.;
  } else
    return 0.;
}
double LogarithmicHaloPotentialPlanarR2deriv(double R,double phi,
					     double t,
					     struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double c= *(args+2); // skip q
  double onem1overb2= *(args+3);
  //Calculate Rforce
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return amp * (1.- 2.*Rt2/(Rt2+c))/(Rt2+c)*Rt2/R/R;
  } else
    return amp * (1.- 2.*R*R/(R*R+c))/(R*R+c);
}
double LogarithmicHaloPotentialPlanarphi2deriv(double R,double phi,
					       double t,
					       struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double c= *(args+2); // skip q
  double onem1overb2= *(args+3);
  //Calculate Rforce
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return - amp * onem1overb2 * (0.5 * pow(R*R*sin(2.*phi),2.) * onem1overb2 \
				  /(Rt2+c)/(Rt2+c)+R*R/(Rt2+c)*cos(2.*phi));
  } else
    return 0.;
}
double LogarithmicHaloPotentialPlanarRphideriv(double R,double phi,
					     double t,
					     struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double c= *(args+2); // skip q
  double onem1overb2= *(args+3);
  //Calculate Rforce
  double Rt2;
  if ( onem1overb2 < 1 ) {
    Rt2= R*R * (1. - onem1overb2 * pow(sin(phi),2));
    return - amp * c / (Rt2+c) / (Rt2+c) * R * sin(2.*phi) * onem1overb2;
  } else
    return 0.;
}
double LogarithmicHaloPotentialDens(double R,double Z, double phi,
				    double t,
				    struct potentialArg * potentialArgs){
  double * args= potentialArgs->args;
  //Get args
  double amp= *args;
  double q= *(args+1);
  double c= *(args+2);
  double onem1overb2= *(args+3);
  //Calculate density
  double zq= Z/q;
  double R2, Rt2, denom, denom2;
  double q2= q*q;
  if ( onem1overb2 < 1 ) {
    R2= R * R;
    Rt2= R2 * (1. - onem1overb2 * pow(sin(phi),2));
    denom= 1. / ( Rt2 + zq * zq + c );
    denom2= denom *  denom;
    return amp * M_1_PI / 4. * ( 2. * Rt2 / R2 * ( denom - Rt2 * denom2 )\
	       + denom / q2 - 2. * zq * zq * denom2 / q2 \
	       - onem1overb2 \
	       * ( 2. * R2 * pow ( sin ( 2. * phi ),2) / 4. * onem1overb2 \
		   * denom2 + denom * cos( 2. * phi ) ) );
  } else
    return amp * M_1_PI / 4. / q2 * ( ( 2. * q2 + 1. ) * c + R * R	\
				       + ( 2. - 1. / q2 ) * Z * Z )/	\
      pow( R * R + zq * zq + c ,2.);
}
